long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -98,17 +101,42 @@ public static void main(String[] args) throws Exception {
     @VisibleForTesting
     public static class AlertFunction extends KeyedProcessFunction<Long, TaxiRide, Long> {
 
+        private ValueState<TaxiRide> taxiRideState;
+
         @Override
         public void open(Configuration config) throws Exception {
-            throw new MissingSolutionException();
+            taxiRideState = getRuntimeContext().getState(new ValueStateDescriptor<>("ride_event", TaxiRide.class));
         }
 
         @Override
-        public void processElement(TaxiRide ride, Context context, Collector<Long> out)
-                throws Exception {}
+        public void processElement(TaxiRide ride, Context context, Collector<Long> out) throws IOException {
+            if(taxiRideState.value() == null){
+                taxiRideState.update(ride);
+                if (ride.isStart) {
+                    context.timerService().registerEventTimeTimer(ride.eventTime.plusSeconds(120 * 60).toEpochMilli());
+                }
+            } else {
+                if (ride.isStart) {
+                    if(Duration.between(ride.eventTime, taxiRideState.value().eventTime).toHours() >= 2)
```

Comment:
I would recommend introduce function (rideTooLong) and avoid having duplicate logic.